task/host-sp-comms/src/main.rs

```diff
@@ -0,0 +1,314 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#![no_std]
+#![no_main]
+
+#[cfg(any(feature = "stm32h743", feature = "stm32h753"))]
+use drv_stm32h7_usart as drv_usart;
+
+use drv_usart::Usart;
+use heapless::Vec;
+use host_sp_messages::{
+    DecodeFailureReason, Header, HostToSp, HubpackError, SpToHost,
+    MAX_MESSAGE_SIZE,
+};
+use mutable_statics::mutable_statics;
+use ringbuf::{ringbuf, ringbuf_entry};
+use userlib::{
+    sys_irq_control, sys_recv_closed, task_slot, TaskId, UnwrapLite,
+};
+
+task_slot!(SYS, sys);
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum UartLog {
+    Tx(u8),
+    TxFull,
+    Rx(u8),
+    RxOverrun,
+}
+
+ringbuf!(UartLog, 64, UartLog::Rx(0));
+
+/// Notification mask for USART IRQ; must match configuration in app.toml.
+const USART_IRQ: u32 = 1;
+
+#[export_name = "main"]
+fn main() -> ! {
+    let uart = configure_uart_device();
+    let tx_buf = mutable_statics! {
+        static mut UART_TX_BUF: [u8; MAX_MESSAGE_SIZE] = [0; _];
+    };
+    let rx_buf = claim_uart_rx_buf();
+
+    sys_irq_control(USART_IRQ, true);
+
+    loop {
+        // Wait for uart interrupt; if we haven't enabled tx interrupts, this
+        // blocks until there's data to receive.
+        let _ = sys_recv_closed(&mut [], USART_IRQ, TaskId::KERNEL);
+
+        // Clear any RX overrun errors. If we hit this, we will likely fail to
+        // decode the next message from the host, which will cause us to send a
+        // `DecodeFailure` response.
+        if uart.check_and_clear_rx_overrun() {
+            ringbuf_entry!(UartLog::RxOverrun);
+        }
+
+        let maybe_response = 'response: loop {
+            // Receive until we find a message delimiter. Since we're using
+            // corncobs for framing, we're looking for 0x00.
+            while let Some(byte) = uart.try_rx_pop() {
+                ringbuf_entry!(UartLog::Rx(byte));
+
+                if byte == 0x00 {
```

Comment:
It seems like this could fail if it turns on midway through the first packet, since we'd be trying to parse a partial `rx_buf`.

We should probably wait for the first `0x00`, _then_ start accumulating data in `rx_buf` and processing it on _subsequent_ `0x00` (either tracked with a boolean flag or a tiny state machine).

Reply:
Trying to parse a partial `rx_buf` ought to be "okay" in that we'll fail to parse and send an error back to the host; failing to respond at all (in this preliminary version) seems like it would be worse. We (cc @citrus-it) still need to figure out how we notify and get sync'd with the host when we start (or restart, in event of a task panic).